protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,638 @@
+                    RequestResponseEvent::Message { peer, message } => match message {
+                        RequestResponseMessage::Request {
+                            request_id: _,
+                            request,
+                            channel,
+                        } => match self.handle_request(peer, request) {
+                            Some(addrs) => {
+                                self.ongoing_inbound.insert(peer, (addrs, channel));
+                                return Poll::Ready(NetworkBehaviourAction::DialPeer {
+                                    peer_id: peer,
+                                    handler: self.inner.new_handler(),
+                                    condition: DialPeerCondition::Always,
+                                });
+                            }
+                            None => {
+                                let response = DialResponse::Err(ResponseError::DialRefused);
+                                let _ = self.inner.send_response(channel, response);
+                            }
+                        },
+                        RequestResponseMessage::Response {
+                            request_id: _,
+                            response,
+                        } => {
+                            if let Some(status) = self.handle_response(peer, Some(response)) {
+                                self.reachability = Reachability::from(&status);
+                                self.last_probe = Instant::now();
+                                return Poll::Ready(NetworkBehaviourAction::GenerateEvent(status));
+                            }
+                        }
+                    },
+                    RequestResponseEvent::ResponseSent { .. } => {}
+                    RequestResponseEvent::OutboundFailure { peer, .. } => {
+                        if let Some(status) = self.handle_response(peer, None) {
+                            self.reachability = Reachability::from(&status);
+                            self.last_probe = Instant::now();
+                            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(status));
+                        }
+                    }
+                    RequestResponseEvent::InboundFailure { peer, .. } => {
+                        self.ongoing_inbound.remove(&peer);
+                    }
+                },
+                Poll::Ready(action) => return Poll::Ready(action.map_out(|_| unreachable!())),
+                Poll::Pending => return Poll::Pending,
+            }
+            if let Some(probe) = self.pending_probe.take() {
+                let records: Vec<_> = params.external_addresses().collect();
+                self.do_probe(probe, records)
+            } else if let Some(auto_retry) = self.auto_retry.as_ref() {
+                if Instant::now().duration_since(self.last_probe) > auto_retry.interval {
```

Comment:
Do I understand correctly that the underlying task is not woken up once `auto_retry.interval` fires? How about using `futures-timer` here, to ensure the timer triggers the underlying logic even if there is no other event calling `poll`?

Reply:
Should be solved now with https://github.com/libp2p/rust-libp2p/pull/2262/commits/99418f85f829a6a00d93c8a530d23e524c233ab0. 
Could you please double check this logic, as I am not too confident with `futures-timer`, and wakers in general?

Reply:
Usage of `futures-timer::Delay` looks good to me. Thank you for the patch @elenaf9.